salt/_modules/metalk8s_drain.py

```diff
@@ -0,0 +1,512 @@
+    def not_found(response):
+        '''Check whether a K8s API response indicates a resource not found.'''
+        return response is None or len(response.items) == 0
+
+    def wait_for_delete(self, pods, interval, timeout, get_pod_fn):
+        '''Wait for pods deletion.'''
+        total_t = 0
+        while total_t < timeout and pods:
+            pending = []
+            loop_s = time.time()    # For time processing pods
+            for pod in pods:
+                response = get_pod_fn(pod.metadata.namespace, pod.metadata.name)
+                if any([
+                        self.not_found(response),
+                        response.items[0].metadata.uid != pod.metadata.uid
+                ]):
+                    log.info("%s evicted", pod.metadata.name)
+                else:
+                    pending.append(pod)
+            loop_d = time.time() - loop_s
+            pods = pending
+            if pods and loop_d < interval:
+                time.sleep(interval - loop_d)
+            total_t += time.time() - loop_s
+        return pods
+
+    def evict_pod(self, pod):
+        '''Evict a pod.'''
+        delete_options = V1DeleteOptions()
+        if self._grace_period >= 0:
+            delete_options.grace_period = self._grace_period
+
+        object_meta = V1ObjectMeta(
+            name=pod.metadata.name,
+            namespace=pod.metadata.namespace)
+
+        eviction = V1beta1Eviction(
+            delete_options=delete_options,
+            metadata=object_meta,
+        )
+
+        api_call = "CoreV1Api->create_namespaced_pod_eviction"
+        api_instance = kubernetes.client.CoreV1Api()
+        try:
+            return api_instance.create_namespaced_pod_eviction(
+                name=eviction.metadata.name,
+                namespace=eviction.metadata.namespace,
+                body=eviction,
+            )
+        except (ApiException, HTTPError) as exc:
+            if isinstance(exc, ApiException) and exc.status == 404:
+                return None
+
+            log.exception('Exception when calling %s', api_call)
+            raise CommandExecutionError(exc)
+
+
+def node_drain(node_name,
+               force=False,
+               grace_period=1,
+               ignore_daemonset=False,
+               timeout=0,
+               delete_local_data=False,
+               dry_run=False,
+               **kwargs):
+    '''
+    Drain the the node identified by the name `node_name`.
+
+    CLI Examples::
+
+        salt '*' kubernetes.node_drain node_name="minikube"
+    '''
+
+    cfg = __salt__['metalk8s_kubernetes.setup_conn'](**kwargs)
+    drainer = Drain(
+        node_name,
+        force=force,
+        grace_period=grace_period,
+        ignore_daemonset=ignore_daemonset,
+        timeout=timeout,
+        delete_local_data=delete_local_data
+    )
+    try:
+        result = drainer.run_drain(dry_run=dry_run, **kwargs)
+    except CommandExecutionError as exc:
+        return False, exc.message
```

Comment:
Is it common practice to return a tuple from a Salt execution function?

Reply:
Nope! In fact, I should either return `None` or let the `CommandExecutionError` fall through.